node/core/proposer/src/lib.rs

```diff
@@ -40,8 +40,12 @@ use sp_transaction_pool::TransactionPool;
 use prometheus_endpoint::Registry as PrometheusRegistry;
 use std::{fmt, pin::Pin, sync::Arc, time};
 
-/// How long proposal can take before we give up and err out
-const PROPOSE_TIMEOUT: core::time::Duration = core::time::Duration::from_millis(2500);
+/// How long proposing can take, before we give up and err out. We need a relatively large timeout
+/// here as long as we have large payload in statement distribution. Assuming we can reach most
+/// nodes within two hops, we will take about 2 seconds for transferring statements (data transfer
+/// only). If necessary, we could be able to reduce this to 3 seconds. To consider: The lower the
+/// riskier that we will not be able to include a candidate.
+const PROPOSE_TIMEOUT: core::time::Duration = core::time::Duration::from_millis(4000);
```

Comment:
What needs to happen after that 4 seconds and can it fit within the remaining two seconds?

Reply:
Invoking the `on_initialize`,  adding transactions, and then `on_finalize` of the block. 2 seconds is a bit of a squeeze, honestly.

When this timeout is hit, it means that no parachain stuff will be included which probably makes the 2 seconds more acceptable.